src/cmd/run.js

```diff
@@ -69,32 +87,83 @@ export default async function run(
 
   const manifestData = await getValidatedManifest(sourceDir);
 
-  const firefoxDesktopRunnerParams = {
+  const runners = [];
+
+  // Options shared between the created runners.
+  const commonRunnerParams = {
     // Common options.
     extensions: [{sourceDir, manifestData}],
     keepProfileChanges,
     startUrl,
     desktopNotifications,
+  };
 
-    // Firefox specific CLI options.
-    firefoxBinary: firefox,
-    profilePath: firefoxProfile,
-    customPrefs,
-    browserConsole,
-    preInstall,
+  if (!target || target.length === 0 || target.includes('firefox-desktop')) {
+    const firefoxDesktopRunnerParams = {
+      ...commonRunnerParams,
 
-    // Firefox runner injected dependencies.
-    firefoxApp,
-    firefoxClient,
-  };
+      // Firefox specific CLI options.
+      firefoxBinary: firefox,
+      profilePath: firefoxProfile,
+      customPrefs,
+      browserConsole,
+      preInstall,
+
+      // Firefox runner injected dependencies.
+      firefoxApp,
+      firefoxClient,
+    };
 
-  const firefoxDesktopRunner = new FirefoxDesktopExtensionRunner(
-    firefoxDesktopRunnerParams
-  );
+    const firefoxDesktopRunner = new FirefoxDesktopExtensionRunner(
+      firefoxDesktopRunnerParams
+    );
+
+    runners.push(firefoxDesktopRunner);
+  }
+
+  if (target && target.includes('firefox-android')) {
+    const firefoxAndroidRunnerParams = {
+      ...commonRunnerParams,
+
+      // Firefox specific CLI options.
+      profilePath: firefoxProfile,
+      customPrefs,
+      browserConsole,
+      preInstall,
+      firefoxApk,
+      adbDevice,
+      adbHost,
+      adbPort,
+      adbBinary,
+
+      // Injected dependencies.
+      firefoxApp,
+      firefoxClient,
+      adb: defaultADB,
+      desktopNotifications: defaultDesktopNotifications,
+      buildSourceDir: (extensionSourceDir: string) => {
+        return buildExtension({
+          sourceDir: extensionSourceDir,
+          artifactsDir,
+          ignoreFiles,
+          asNeeded: false,
+          // TODO: choose a different artifactsDir for additional safety?
```

Comment:
@kumar303 tons of unit tests (and also some "code removal") later... I think that PR is almost ready for you.

Unfortunately even after the "code removal" phase, it is still a quite big (most of it is composed by the new extension runner file and its related test file).
On the bright side, the changes in this PR are enough to be able to also use the feature for real and being sure that the unit tests are testing actual real scenarios
(I'm still tempted to add some additional test case :-P).

This line is the only remaining one from my current todo list related to this PR that I haven't worked on yet, we can probably defer it to a follow up PR (and prevent this one from growing further).

I'll take another look at the final diff of this PR tomorrow, and then we can start to review it.